diff --git a/CMakeLists.txt b/CMakeLists.txt
index eb01a8828..561e8cdc3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -616,6 +616,7 @@ add_subdirectory(PWCOND)
 add_subdirectory(TDDFPT)
 add_subdirectory(XSpectra)
 add_subdirectory(QEHeat)
+add_subdirectory(D3Q)
 if(QE_ENABLE_DOC)
     add_subdirectory(Doc)
 endif()
@@ -837,3 +838,10 @@ add_custom_target(all_currents
 	qe_qeheat_exe
     COMMENT
 	"QEHeat code to compute energy and electronic density currents")
+
+add_custom_target(d3q
+    DEPENDS
+        qe_d3q_exe
+        qe_d3_thermal2
+    COMMENT
+        "D3Q code to compute third derivative of total energy, and tools to compute phonon linewidth and thermal conductivity")
